File: Frontend/angular-app/src/app/modules/commodity/repositories/commodity-category/commands/create-commodity-category-command.ts
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../core/models/service-result";
import {HttpService} from "../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {CommodityCategory} from "../../../entities/commodity-category";
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";

export class CreateCommodityCategoryCommand extends IRequest<CreateCommodityCategoryCommand, CommodityCategory> {

  public parentId: number | undefined = undefined;
  public levelCode: string | undefined = undefined;
  public code: string | undefined = undefined;
  public codingMode: number | undefined = undefined;
  public title: string | undefined = undefined;
  public measureId: number | undefined = undefined;
  public orderIndex: number | undefined = undefined;
  public isReadOnly: boolean | undefined = undefined;
  public requireParentProduct: boolean | undefined = undefined;

  constructor() {
    super();
  }

  mapFrom(entity: CommodityCategory): CreateCommodityCategoryCommand {
    throw new ApplicationError(CreateCommodityCategoryCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  mapTo(): CommodityCategory {
    throw new ApplicationError(CreateCommodityCategoryCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/commodity/commodityCategory/add";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(CreateCommodityCategoryCommandHandler.name)
export class CreateCommodityCategoryCommandHandler implements IRequestHandler<CreateCommodityCategoryCommand, CommodityCategory> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService,
  ) {
  }

  async Handle(request: CreateCommodityCategoryCommand): Promise<CommodityCategory> {
    let httpRequest: HttpRequest<CreateCommodityCategoryCommand> = new HttpRequest<CreateCommodityCategoryCommand>(request.url, request);
    // httpRequest.Query  += ;
    // httpRequest.Headers = httpRequest.Headers.append()

    return await this._httpService.Post<CreateCommodityCategoryCommand, ServiceResult<CommodityCategory>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })


  }
}

File: Backend/Microservices/Commodity/Eefa.Commodity.Application/Commands/CommodityCategory/Create/CreateCommodityCategoryCommand.cs
using AutoMapper;
using Eefa.Common.CommandQuery;
using Eefa.Common;
using Eefa.Common.Data;
using MediatR;
using System.Threading;
using System.Threading.Tasks;
using Eefa.Commodity.Application.Queries.Commodity;

namespace Eefa.Commodity.Application.Commands.CommodityCategory.Create
{
    public class CreateCommodityCategoryCommand : CommandBase, IRequest<ServiceResult<CommodityCategoryModel>>, IMapFrom<CreateCommodityCategoryCommand>, ICommand
    {

        /// <summary>
        /// کد والد
        /// </summary>
        public int? ParentId { get; set; }

        /// <summary>
        /// کد سطح
        /// </summary>
        public string LevelCode { get; set; } = default!;
        /// <summary>
        /// کد
        /// </summary>
        public string Code { get; set; } = default!;
        public int CodingMode { get; set; } = default!;

        /// <summary>
        /// نام اختصاصی
        /// </summary>
        public string? UniqueName { get; set; }

        /// <summary>
        /// عنوان
        /// </summary>
        public string? Title { get; set; }

        /// <summary>
        /// کدواحد اندازه گیری
        /// </summary>
        public int? MeasureId { get; set; }

        /// <summary>
        /// ترتیب نمایش
        /// </summary>
        public int OrderIndex { get; set; } = default!;

        /// <summary>
        /// آیا فقط قابل خواندن است؟
        /// </summary>
        public bool IsReadOnly { get; set; } = default!;

        public bool? RequireParentProduct { get; set; }


        public void Mapping(Profile profile)
        {
            profile.CreateMap<CreateCommodityCategoryCommand, Data.Entities.CommodityCategory>()
                .IgnoreAllNonExisting();

        }
    }

    public class CreateCommodityCategoryCommandHandler : IRequestHandler<CreateCommodityCategoryCommand, ServiceResult<CommodityCategoryModel>>
    {
        private readonly IRepository<Data.Entities.CommodityCategory> _repository;
        private readonly IMapper _mapper;

        public CreateCommodityCategoryCommandHandler(IRepository<Data.Entities.CommodityCategory> repository, IMapper mapper)
        {
            _mapper = mapper;
            _repository = repository;
        }


        public async Task<ServiceResult<CommodityCategoryModel>> Handle(CreateCommodityCategoryCommand request, CancellationToken cancellationToken)
        {
            var commodityCategory = _mapper.Map<Data.Entities.CommodityCategory>(request);
            request.MeasureId = request.MeasureId == 0 ? null : request.MeasureId;

            var entity = _repository.Insert(commodityCategory);

            await _repository.SaveChangesAsync(cancellationToken);

            return ServiceResult<CommodityCategoryModel>.Success(
                _mapper.Map<CommodityCategoryModel>(entity)
            );
        }
    }
}
